jib-core/src/main/java/com/google/cloud/tools/jib/frontend/JavaLayerConfigurations.java

```diff
@@ -115,14 +117,49 @@ public Builder addDirectoryContents(
         Predicate<Path> pathFilter,
         AbsoluteUnixPath basePathInContainer)
         throws IOException {
+      return addDirectoryContents(
+          layerType, sourceRoot, pathFilter, basePathInContainer, ImmutableMap.of());
+    }
+
+    /**
+     * Adds directory contents to a layer selectively (via {@code pathFilter}) and recursively.
+     * {@code sourceRoot} must be a directory. Empty directories will always be added regardless of
+     * {@code pathFilter}, except that {@code sourceRoot} is never added. Permissions are specified
+     * via {@code permissionsMap}, which maps from extraction path on the container to file
+     * permissions.
+     *
+     * @param layerType the layer to add files into
+     * @param sourceRoot root directory whose contents will be added
+     * @param pathFilter filter that determines which files (not directories) should be added
+     * @param basePathInContainer directory in the layer into which the source contents are added
+     * @param permissionsMap the map from absolute path on container to file permission
+     * @return this
+     * @throws IOException error while listing directories
+     * @throws NotDirectoryException if {@code sourceRoot} is not a directory
+     */
+    // TODO: Use in plugins
+    public Builder addDirectoryContents(
```

Comment:
I think we need the permission version for `addFile()` too (the single-file addition method). It's being used.